src/errors.rs

```diff
@@ -1,67 +1,154 @@
-use output;
-use std::ffi::OsString;
-
-const ERROR_PREFIX: &'static str = "CLI assertion failed";
-
-fn format_cmd(cmd: &[OsString]) -> String {
-    let result: Vec<String> = cmd.iter()
-        .map(|s| s.to_string_lossy().into_owned())
-        .collect();
-    result.join(" ")
-}
-
-error_chain! {
-    links {
-        Output(output::Error, output::ErrorKind);
-    }
-    foreign_links {
-        Io(::std::io::Error);
-        Fmt(::std::fmt::Error);
-    }
-    errors {
-        SpawnFailed(cmd: Vec<OsString>) {
-            description("Spawn failed")
-            display(
-                "{}: (command `{}` failed to run)",
-                ERROR_PREFIX,
-                format_cmd(cmd),
-            )
-        }
-        AssertionFailed(cmd: Vec<OsString>) {
-            description("Assertion failed")
-            display(
-                "{}: (command `{}` failed)",
-                ERROR_PREFIX,
-                format_cmd(cmd),
-            )
-        }
-        StatusMismatch(expected: bool, out: String, err: String) {
-            description("Wrong status")
-            display(
-                "Expected to {}\nstatus={}\nstdout=```{}```\nstderr=```{}```",
-                expected = if *expected { "succeed" } else { "fail" },
-                got = if *expected { "failed" } else { "succeeded" },
-                out = out,
-                err = err,
-            )
-        }
-        ExitCodeMismatch(
-            expected: Option<i32>,
-            got: Option<i32>,
-            out: String,
-            err: String
-        ) {
-            description("Wrong exit code")
-            display(
-                "Expected exit code to be `{expected:?}`)\n\
-                exit code=`{code:?}`\n\
-                stdout=```{stdout}```\n\
-                stderr=```{stderr}```",
-                expected=expected,
-                code=got,
-                stdout=out,
-                stderr=err,
-            )
+use std::fmt;
+use std::result;
+
+use failure;
+
+pub type Result<T> = result::Result<T, failure::Error>;
```

Comment:
I think this is the crux of this PR: We should probably use a concrete error type here and `impl From` a bunch of times! This way, we have way more control over the way we handle errors. We can still do `type Result<T> = StdResult<T, Box<AssertCliError>` to make the type a bit smaller in memory, but I think we lose a lot of options if we only deal with a trait object here.